lib/network/discovery_message.go

```diff
@@ -0,0 +1,170 @@
+package network
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/btcsuite/btcutil/base58"
+
+	"boscoin.io/sebak/lib/common"
+	"boscoin.io/sebak/lib/common/keypair"
+	"boscoin.io/sebak/lib/errors"
+	"boscoin.io/sebak/lib/node"
+)
+
+type DiscoveryMessage struct {
+	H DiscoveryMessageHeader
+	B DiscoveryMessageBody
+}
+
+type DiscoveryMessageHeader struct {
+	Signature string `json:"signature"`
+}
+
+type DiscoveryMessageBody struct {
+	Created    string            `json:"created"`
+	Address    string            `json:"address"`  // LocalNode.Address()
+	Endpoint   *common.Endpoint  `json:"endpoint"` // LocalNode.publishEndpoint()
+	Validators []*node.Validator `json:"validators"`
+}
+
+func NewDiscoveryMessage(localNode *node.LocalNode, validators ...*node.Validator) (dm DiscoveryMessage, err error) {
+	// `PublishEndpoint()` must be not empty
+	if localNode.PublishEndpoint() == nil {
+		err = errors.EndpointNotFound
+		return
+	}
+
+	dm = DiscoveryMessage{
+		H: DiscoveryMessageHeader{},
+		B: DiscoveryMessageBody{
+			Endpoint:   localNode.PublishEndpoint(),
+			Validators: validators,
+		},
+	}
+
+	return
+}
+
+func DiscoveryMessageFromJSON(b []byte) (dm DiscoveryMessage, err error) {
+	err = json.Unmarshal(b, &dm)
+	return
+}
+
+func (db DiscoveryMessageBody) MakeHashString() string {
+	return base58.Encode(common.MustMakeObjectHash(db))
+}
+
+func (dm DiscoveryMessage) GetHash() string {
+	return dm.B.MakeHashString()
+}
+
+func (dm DiscoveryMessage) Equal(common.Message) bool {
+	return false
+}
+
+func (dm DiscoveryMessage) Source() string {
+	return dm.B.Address
+}
+
+func (dm DiscoveryMessage) Version() string {
+	return common.DiscoveryVersionV1
+}
+
+func (dm DiscoveryMessage) GetType() common.MessageType {
+	return common.DiscoveryMessage
+}
+
+func (dm DiscoveryMessage) IsWellFormed(conf common.Config) error {
+	if len(dm.H.Signature) < 1 {
+		return errors.InvalidMessage
+	}
+	if len(dm.B.Created) < 1 {
+		return errors.InvalidMessage
+	}
+	if len(dm.B.Address) < 1 {
+		return errors.InvalidMessage
+	}
+	if len(dm.B.Endpoint.String()) < 1 {
+		return errors.InvalidMessage
+	}
+
+	// check time
+	created, err := common.ParseISO8601(dm.B.Created)
+	if err != nil {
+		return err
+	}
+	now := time.Now()
+	sub := created.Sub(now)
+	if sub < 0 && sub < (common.DiscoveryMessageCreatedAllowDuration*-1) {
+		return errors.MessageHasIncorrectTime
+	}
+
+	if sub > 0 && sub > common.DiscoveryMessageCreatedAllowDuration {
```

Comment:
`common.DiscoveryMessageCreatedAllowDuration` is `10 sec`, so in this condition `sub > 0 && sub > 10`, `sub > 0` is redundant.